### ISSUE TITLE
Checks should be independent from one another

### ISSUE THREAD

**Issue by m4rc1e** [googlefonts/fontbakery#1292]
Several checks rely on each other to produce the right result. This means the order of them is critical. Every check should be isolatable.

I've noticed this for the following tests: https://github.com/googlefonts/fontbakery/blob/779b31c5cb363e38d581c6dc6dd435ad09fc3176/fontbakery-check-ttf.py#L217-L224

for the fsSelection checks, I'd consolidate them into a single check. The same could be applied to the macStyle checks.

**Comment by felipesanches** [comment]
Do you want to keep this opened as a general principle of eliminating all and every instance of coupled checks? In some places we do things like "check if license files exist and error in case they don't". Later another check is skipped unless the license files are there. Re-checking (just for the sake of making the checks fully self-contained) would end up being a bit redundant. I'm not sure if I want to put much energy on this kind of code restructuring. What do you think, @m4rc1e ?

**Comment by felipesanches** [comment]
I'm refering to this:

https://github.com/googlefonts/fontbakery/blob/17ad3512f894414b65a343dded9c108279d670cf/fontbakery-check-ttf.py#L224-L226

```python
    found = checks.check_font_has_a_license(fb, file_path)
    checks.check_copyright_entries_match_license(fb, found, file_path, font)
    checks.check_font_has_a_valid_license_url(fb, found, font)
```

**Comment by graphicore** [comment]
I'm thinking it could be modeled using a dependency graph approach. But, that could also be overkill for this example.
The thought is, that `check_copyright_entries_match_license` and `check_font_has_a_valid_license_url` would state explicitly that they depend on the result of `checks.check_font_has_a_license` being positiv, i.e. that the check passes or maybe they require result/return value of the check function.
It would need a nice way to describe that dependency and it would only make sense on the road to a rather smart and customizable font testing suite. For example, some tests would be able to pull tests they depend on automatically, even if these dependencies haven't been configured explicitly.

**Comment by graphicore** [comment]
> Do you want to keep this opened as a general principle of eliminating all and every instance of coupled checks?

Keep it open, it's at least an issue we should solve in some way or another :-).

**Comment by felipesanches** [comment]
Yes, I had the same thoughts about using a dependency-graph to implement this. But then I also ended up quickly concluding it would be too fancy and overkill.

**Comment by graphicore** [comment]
Well, if you think about making this a useful tool for other parties as well, who will probably need to test against their own custom spec. It may become a nice feature to model your own spec, without having to call each test explicitly.

**Comment by davelab6** [comment]
I worry you are reimplementing existing thorough testing frameworks. Maybe
one of you should do a review of popular ones and pick one and migrate to
it.

I don't want anyone here writing a dependency graph engine. Yak.

**Comment by felipesanches** [comment]
We should also consider here this other request from @davelab6 on #971 : 

> each fix line should set a fix flag, and
> if that flag is True by this point, only then write the file
> and then say any further output regards fixed files, and
> re-run the script on each fixed file with logging level = error
> so no info-level log items are shown

I think it is closely related. It may be good to think about both issues at once, in order to find a good solution to both.

**Comment by felipesanches** [comment]
What I mean is that one possible approach to implement #971 is by running the full set of checks performing hotfixes and then re-run everything again on the resulting hotfixed TTF.

The actual hotfixes performed, though, would potentially depend on the order in which the checks are invoked.

**Comment by davelab6** [comment]
Moving to 0.2.0 and marking as important :)

**Comment by felipesanches** [comment]
This has been fixed on our new architecture (that is already available on master and will be released for the first time on 0.3.0)
